transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu

```diff
@@ -46,1446 +45,608 @@
-
-    int64_t o_dim[4] =  {b, h, s_q, d};
-    int64_t o_stride[4];
-    generateMatrixStrides(b, h, s_q, s_kv, d, o_stride, layout, NVTE_QKV_Matrix::NVTE_O_Matrix);
-
-    int64_t seqlen_dim[4]    = {b, 1, 1, 1};
-    int64_t seqlen_stride[4] = {1, 1, 1, 1};
-
-    auto seqlenQTensor = tensor_create(
-        CUDNN_DATA_INT32, Q_SEQLEN_ID, seqlen_dim, seqlen_stride, false, false);
-    auto seqlenKTensor = tensor_create(
-        CUDNN_DATA_INT32, K_SEQLEN_ID, seqlen_dim, seqlen_stride, false, false);
-
-    auto vTensor = tensor_create(tensorType, V_ID, v_dim, v_stride, false, false);
-    // second GEMM output
-    auto oTensor = tensor_create(tensorType, O_ID, o_dim, o_stride, false, false);
-
-    // Define the matmul 2 desc
-    auto matmul_2_Desc = cudnn_frontend::MatMulDescBuilder()
-                            .setComputeType(CUDNN_DATA_FLOAT)
-                            .setPaddingValue(0.0f)
-                            .build();
-
-    // Create a matmul 2 node
-    auto&& matmul_op_builder =
-        cudnn_frontend::OperationBuilder(CUDNN_BACKEND_OPERATION_MATMUL_DESCRIPTOR);
-
-    matmul_op_builder.setaMatDesc(afterScaleDropoutTensor)
-                     .setbMatDesc(vTensor)
-                     .setcMatDesc(oTensor)
-                     .setmatmulDesc(matmul_2_Desc);
-
-    if (padding_aware) {
-        matmul_op_builder.setmOverrideDesc(seqlenQTensor).setkOverrideDesc(seqlenKTensor);
+void fused_attn_arbitrary_seqlen_fwd_impl(
+                int64_t b, int64_t h, int64_t hg, int64_t s_q, int64_t s_kv, int64_t d,
+                bool is_training, float scaling_factor, float dropout_probability,
+                NVTE_QKV_Layout layout,
+                NVTE_Bias_Type bias_type, NVTE_Mask_Type mask_type,
+                void *devPtrQ, void *devPtrK, void *devPtrV, void *devPtrBias,
+                void *devPtrSoftmaxStats, void *devPtrO,
+                void* devPtrDropoutSeed, void* devPtrDropoutOffset,
+                void* devPtrCuSeqlensQ, void* devPtrCuSeqlensKV,
+                cudnn_frontend::DataType_t tensorType,
+                void *workspace, size_t *workspace_size,
+                cudaStream_t stream, cudnnHandle_t handle, bool* check_support) {
+    NVTE_CHECK_CUDNN(cudnnSetStream(handle, stream));
+
+    bool is_bias = (bias_type == NVTE_Bias_Type::NVTE_POST_SCALE_BIAS);
+    bool is_alibi = (bias_type == NVTE_Bias_Type::NVTE_ALIBI);
+    bool is_causal = ((mask_type == NVTE_Mask_Type::NVTE_CAUSAL_MASK)
+        || (mask_type == NVTE_Mask_Type::NVTE_PADDING_CAUSAL_MASK));
+    bool is_padding = ((mask_type == NVTE_Mask_Type::NVTE_PADDING_MASK)
+        || (mask_type == NVTE_Mask_Type::NVTE_PADDING_CAUSAL_MASK));
+    bool is_dropout = (dropout_probability != 0.0f);
+
+    // provide a mechanism for support checking for upcoming FE APIs
+    bool tmp_check = *check_support;
+    if (*check_support && !tmp_check) {
+        *check_support = tmp_check;
+        return;
```

Comment:
Could you provide more insight on what is this used to? I can't figure out in which situation will go into the branch of line 72.

Reply:
I was thinking of replacing the content of `nvte_get_fused_attn_backend` with `graph.check_support()` from each backend, so we don't need to maintain the condition checks in `nvte_get_fused_attn_backend` manually. The `check_support` flag can be used as a way to say 'check support please' and then also return the check result. Not the cleanest design. Will have to think about this. Also, v1/prerelease 4's `check_support()` API doesn't do all the checks we currently do yet, so I might just remove this feature for now.